enhancements/authentication/synchronizing-user-groups-on-login.md

```diff
@@ -0,0 +1,238 @@
+---
+title: synchronizing-user-groups-on-login
+authors:
+  - "@stlaz"
+reviewers:
+  - "@sur"
+  - "@slaskawi"
+approvers:
+  - "@deads2k"
+  - "@sttts"
+creation-date: 2021-05-22
+last-updated: yyyy-mm-dd
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Synchronizing User Groups on Login
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OpenShift should be capable of synchronizing user groups from a 3rd party
+identity provider upon a user logging in to their OpenShift cluster.
+
+## Motivation
+
+In the real world, OpenShift is not the owner of user identities but these
+are stored in a 3rd party identity provider system. When OpenShift retrieves
+such an identity, it usually only gathers information about the user's unique
+identification in that IdP and optionally a username that's preferred for the
+user if it's different from its identifier.
+
+It is a common practice to group users in logical groups on the identity provider
```

Comment:
The most generic spec we could leverage on is OpenID Connect. Typically, groups information might be put in the Access Token (from oAuth), ID Token (OpenID Connect) or we may grab it from User Info Endpoint (also OpenID Connect). Here are some examples from my local Keycloak instance:

* [Access Token with Groups Mapper turned on](https://jwt.io/#debugger-io?token=eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJmZkVnekFYUnR3bXpvU0xJWXJCbXJ4VDBfVkQ2bmMzV1AtaDRJM05HTXRnIn0.eyJleHAiOjE2MjI0NDQ3NzksImlhdCI6MTYyMjQ0NDcxOSwianRpIjoiOWQyZjA4ZTAtNzY2Ni00YTZjLTk2OWYtZTdhOTM2NDIzNWE4IiwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo4MDgwL2F1dGgvcmVhbG1zL21hc3RlciIsImF1ZCI6WyJtYXN0ZXItcmVhbG0iLCJhY2NvdW50Il0sInN1YiI6IjYwMWY5N2NjLTQ2NzgtNGY2Ny04ZTNhLWIwZmZmMmRhZDllNiIsInR5cCI6IkJlYXJlciIsImF6cCI6ImFkbWluLWNsaSIsInNlc3Npb25fc3RhdGUiOiIyYmE2MzZhNy02ZTA5LTQ5OTgtYWFlNS0yNmU5NjU2ZWM4MDIiLCJhY3IiOiIxIiwicmVhbG1fYWNjZXNzIjp7InJvbGVzIjpbImNyZWF0ZS1yZWFsbSIsImRlZmF1bHQtcm9sZXMtbWFzdGVyIiwib2ZmbGluZV9hY2Nlc3MiLCJhZG1pbiIsInVtYV9hdXRob3JpemF0aW9uIl19LCJyZXNvdXJjZV9hY2Nlc3MiOnsibWFzdGVyLXJlYWxtIjp7InJvbGVzIjpbInZpZXctaWRlbnRpdHktcHJvdmlkZXJzIiwidmlldy1yZWFsbSIsIm1hbmFnZS1pZGVudGl0eS1wcm92aWRlcnMiLCJpbXBlcnNvbmF0aW9uIiwiY3JlYXRlLWNsaWVudCIsIm1hbmFnZS11c2VycyIsInF1ZXJ5LXJlYWxtcyIsInZpZXctYXV0aG9yaXphdGlvbiIsInF1ZXJ5LWNsaWVudHMiLCJxdWVyeS11c2VycyIsIm1hbmFnZS1ldmVudHMiLCJtYW5hZ2UtcmVhbG0iLCJ2aWV3LWV2ZW50cyIsInZpZXctdXNlcnMiLCJ2aWV3LWNsaWVudHMiLCJtYW5hZ2UtYXV0aG9yaXphdGlvbiIsIm1hbmFnZS1jbGllbnRzIiwicXVlcnktZ3JvdXBzIl19LCJhY2NvdW50Ijp7InJvbGVzIjpbIm1hbmFnZS1hY2NvdW50IiwibWFuYWdlLWFjY291bnQtbGlua3MiLCJ2aWV3LXByb2ZpbGUiXX19LCJzY29wZSI6Im9wZW5pZCBlbWFpbCBwcm9maWxlIiwiZW1haWxfdmVyaWZpZWQiOmZhbHNlLCJncm91cHMiOlsiY3JlYXRlLXJlYWxtIiwiZGVmYXVsdC1yb2xlcy1tYXN0ZXIiLCJvZmZsaW5lX2FjY2VzcyIsImFkbWluIiwidW1hX2F1dGhvcml6YXRpb24iXSwicHJlZmVycmVkX3VzZXJuYW1lIjoiYWRtaW4ifQ.XjJWgzw17mZOiLBXclWTCjfw1XE9xv9D6s0N0TGNzqihsXzDxZ6Y1vDovIuMYHqtk3-Vs6Ro53_gLF3BDpx8sc66cYOriCjlQGB-eYQuXMFmwa3vzM2I3I_TGXZIJAkEhV9-J1Zr5_DbluHYbb2CaNnkvSZcNBTdR-1DHBIbG4An5LanGkIuPqA_WDSb8H264TgJfM96SFDHkOz-kSuFkKVV_Pm6-oI6OzVl_Xq2Cp2UEhcva-8bdADdOWNkLa8O9KBL_fr9wL2zdv3iULWOMx2aQb6rcS_j5u7v9EzL9S9U2HDK6b-fzwkwJVgidJG0tmfpNe0oRbRQ4bpB5k8pAA&publicKey=-----BEGIN%20PUBLIC%20KEY-----%0AMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAshFo9sR3JdjbxtCzhuXD%0ANbZKZH%2FjaDT6BBJpiSK%2BvVMotGdooDTu1HYPiEwTQ6F1cj6AULiyRB2%2FXjoUEa7S%0AKqoCjG1S9dGaPPp2EccbI7qjtuQDn9J10JErPVPI6vMiH0jX%2BwzYZEibfoNGeK05%0AJ5BZ1sonhmtXQ8lPIpsI8anAYo7yzuKOgKsW2ABjAONtoUUJO7YElRSXtrtqtRAi%0A4tlwAdZOSFDZW4nMY5YIcnzuu29WYsbkEsBNWU8VTAZG3r8UoanyaUIXzEC9iRJr%0AQnWy65eceIzofZP0sDN1L8MldmmytvugbP6hPiMqYc2yGGiPI2yj%2BIQ6mUu%2FoVz%2F%0A%2FQIDAQAB%0A-----END%20PUBLIC%20KEY-----%0A)
* [ID Token obtained with "scope=openid"](https://jwt.io/#debugger-io?token=eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJmZkVnekFYUnR3bXpvU0xJWXJCbXJ4VDBfVkQ2bmMzV1AtaDRJM05HTXRnIn0.eyJleHAiOjE2MjI0NDQ3NzksImlhdCI6MTYyMjQ0NDcxOSwiYXV0aF90aW1lIjowLCJqdGkiOiIyYzA1ZjdhNi0wY2VkLTRiN2UtYTE5Ny05OTlmYTlmZmM4N2EiLCJpc3MiOiJodHRwOi8vbG9jYWxob3N0OjgwODAvYXV0aC9yZWFsbXMvbWFzdGVyIiwiYXVkIjoiYWRtaW4tY2xpIiwic3ViIjoiNjAxZjk3Y2MtNDY3OC00ZjY3LThlM2EtYjBmZmYyZGFkOWU2IiwidHlwIjoiSUQiLCJhenAiOiJhZG1pbi1jbGkiLCJzZXNzaW9uX3N0YXRlIjoiMmJhNjM2YTctNmUwOS00OTk4LWFhZTUtMjZlOTY1NmVjODAyIiwiYXRfaGFzaCI6Ink4ekY3U2VDNXhzaTVKN3lhdmFRcXciLCJhY3IiOiIxIiwiZW1haWxfdmVyaWZpZWQiOmZhbHNlLCJncm91cHMiOlsiY3JlYXRlLXJlYWxtIiwiZGVmYXVsdC1yb2xlcy1tYXN0ZXIiLCJvZmZsaW5lX2FjY2VzcyIsImFkbWluIiwidW1hX2F1dGhvcml6YXRpb24iXSwicHJlZmVycmVkX3VzZXJuYW1lIjoiYWRtaW4ifQ.mWQMJ2YlyEcM0muhvKm71EXBwN5yjL15uYh-3tJetjsWElMV6MKHGKquMx426hkq-ojpWG5fhKhT4YTkiHNEHzWKg4K0pvb6vLabddrQiMpHAzZgIMogKDQn7PVsNxH03Fsepy7mk-WT8VJm9af0enGbDGiQRrV1SGUmGYCWqF5xJDIpgtD3mwTwpQflCZkoZywBslrQQ7JbI9s6Db_MrljAgSUqTQBz_j4mKSFWe2RVf1Pp8YHCXw40F69fGftCnNCZeooZuYvCNs7x_Yg1nEWHpRoi1oQvrjd5zusV14MINOQuF3WYSSojhP0WEAiAioOtHZ96p2fhF9gcuiFsWg&publicKey=-----BEGIN%20PUBLIC%20KEY-----%0AMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAshFo9sR3JdjbxtCzhuXD%0ANbZKZH%2FjaDT6BBJpiSK%2BvVMotGdooDTu1HYPiEwTQ6F1cj6AULiyRB2%2FXjoUEa7S%0AKqoCjG1S9dGaPPp2EccbI7qjtuQDn9J10JErPVPI6vMiH0jX%2BwzYZEibfoNGeK05%0AJ5BZ1sonhmtXQ8lPIpsI8anAYo7yzuKOgKsW2ABjAONtoUUJO7YElRSXtrtqtRAi%0A4tlwAdZOSFDZW4nMY5YIcnzuu29WYsbkEsBNWU8VTAZG3r8UoanyaUIXzEC9iRJr%0AQnWy65eceIzofZP0sDN1L8MldmmytvugbP6hPiMqYc2yGGiPI2yj%2BIQ6mUu%2FoVz%2F%0A%2FQIDAQAB%0A-----END%20PUBLIC%20KEY-----%0A)
* [User Info Endpoint](https://gist.github.com/slaskawi/647d903186f7af4c2a390ce9cd620e3d)

The biggest problem I can see is that there's no standard "groups" claim (see the [spec](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-json-web-token)), which means there's is no guarantee any of those will expose groups in a standard way. We'll rely on vendor-specific behavior here.

So how about implementing a mechanism that will scan the Access and the ID Tokens and try to sync groups automatically - without the additional round-trip to the User Info endpoint?

Reply:
> typically, groups information might be put in the Access Token (from oAuth)
I don't think that's a very typical use, access tokens are usually (as described in the RFC) just opaque strings, although some are trying to put some semantics to it: https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/

> So how about implementing a mechanism that will scan the Access and the ID Tokens and try to sync groups automatically - without the additional round-trip to the User Info endpoint?
yes, the the introspection of the ID token is what we'll do in the implementation, but we won't be touching the access tokens at all

Reply:
Ok, makes sense!